### PR TITLE
Sparse rate matrix

### PR DESCRIPTION
**Pull request by TimonKapischke**
- implemented a sparse rate matrix
- let s = u**t the alphabet size with tuple length t and input alphabet size u (e.g. u=20 for amino acids) than the sparse rate matrix has only (u-1) * t * s/2 parameters instead of (s-1)*s/2
- works for amino acids and nucleotides

### CODE REVIEW

File: tcmc/tcmc.py [hunk 1]
```diff
@@ -100,20 +104,35 @@ def __parse_forest(self,forest):
...
-        self.R_inv = self.add_weight(shape = (M, int(s*(s-1)/2)), name = "R_inv", dtype = tf.float64,
-                                     initializer = rates_initializer)
-        
+        if not self.sparse_rates:
+            # The parameters that we want to learn
+            self.R_inv = self.add_weight(shape = (M, int(s*(s-1)/2)), name = "R_inv", dtype = tf.float64,
+                                         initializer = rates_initializer)
+        else:
+            max_tuple_length = 10
+            nuc_s = [4 ** i for i in range(1, max_tuple_length)]
```
Comment (MarioStanke) [inline tcmc/tcmc.py]: add some comment here, like currently for dna (4) and amino acids (20) alphabets only
Is t initialized somewhere?
What happens when
 1. the tuple length is larger than `max_tuple_length`? - at least a comment
 2. when `s` is not found in either set?

File: tcmc/tensor_utils.py [hunk 2]
```diff
@@ -120,7 +121,60 @@ def batched_segment_indices(segment_lengths):
     
     return tf.stack((seg_ids, batch_indices), axis=1)
 
+def sparse_rate_matrix(M, s):
+    """
+    Returns indices for the upper triangle rate matrix. Allowed are only transitions with maximal one mutation.
```
Comment (MarioStanke) [inline tcmc/tensor_utils.py]:
```suggestion
    Returns indices for the upper triangle rate matrix. Allowed are only transitions with at most one mutation per tuple.
```

File: tcmc/tensor_utils.py [hunk 3]
```diff
@@ -120,7 +121,60 @@ def batched_segment_indices(segment_lengths):
...
+        M (int): Number of models
+        s (int): size of the alphabet
+
+    Returns:
+        tf.Tensor: Tensor with all indices of the trainable parameter for the rate matrix construction
+        tf.Tensor: Tensor with all indices that are not trainable parameter for the rate matrix construction
+    """
+
+    max_tuple_length = 10
+    nuc_alphabet_s = [4 ** i for i in range(2, max_tuple_length)]
```
Comment (MarioStanke) [inline tcmc/tensor_utils.py]: 2 -> min_tuple_length
then the 2 in line 142 will be easier to understand

**Review by MarioStanke — CHANGES_REQUESTED** [review]
Looks good, but please add a few comments and cases, where requested.

File: tcmc/tcmc.py (hunk 1 above)
Comment (TimonKapischke) [reply tcmc/tcmc.py]: Is t initialized somewhere? - No. Only s is initialized.

**Review by MarioStanke — APPROVED** [review]

**Pull request merged**